aten/src/ATen/mps/MPSDevice.mm

```diff
@@ -3,26 +3,72 @@
 #include <c10/util/CallOnce.h>
 
 #include <ATen/mps/MPSDevice.h>
+#include <ATen/mps/IndexKernels.h>
 
 namespace at {
 namespace mps {
 
 static std::unique_ptr<MPSDevice> mps_device;
 static c10::once_flag mpsdev_init;
 
+static inline MTLLanguageVersion getMetalLanguageVersion(const id<MTLDevice>& device) {
+  // MPS Advanced Indexing needs at least Metal 2.0 (support for Argument Buffers and function constants)
+  MTLLanguageVersion languageVersion;
+
+#if defined(__MAC_13_0) && __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_13_0
+  languageVersion = MTLLanguageVersion3_0;
+#elif defined(__MAC_12_0) && __MAC_OS_X_VERSION_MIN_REQUIRED >= __MAC_12_0
+  languageVersion = MTLLanguageVersion2_4;
+#elif
+  #error "Metal is not available on the current platform."
+#endif
+
+  TORCH_CHECK([device supportsFamily:MTLGPUFamilyMac2], "Missing Metal support for MTLGPUFamilyMac2");
+  return languageVersion;
+}
+
 MPSDevice* MPSDevice::getInstance() {
   c10::call_once(mpsdev_init, [] {
       mps_device = std::unique_ptr<MPSDevice>(new MPSDevice());
   });
   return mps_device.get();
 }
 
+id<MTLFunction> MPSDevice::metalIndexingFunction(const std::string& kernel, MTLFunctionConstantValues* constantValues) {
+  assert(_mtl_device);
```

Comment:
Please never use asserts like that, as they are no-ops in release builds. If you want a debug specific assert, please use `TORCH_INTERNAL_ASSERT_DEBUG_ONLY`

Reply:
Done.